gpflow/models/training_mixins.py

```diff
@@ -0,0 +1,58 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar, Union
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData, OutputData)
+
+
+class InternalDataTrainingLossMixin:
+    def training_loss(self) -> tf.Tensor:
+        """
+        Training loss for models that encapsulate their data.
```

Comment:
'encapsulate' isn't quite the right word here. it doesn't have to _hide_ the data, which is implied by 'encapsulation'. It just has to have access to their data by other means

Reply:
Please suggest a better way of phrasing it.

Reply:
```
class InternalDataTrainingLossMixin:
    """
    Mixin to provide convenience methods for training loss. Subclasses will typically own their own
    data (this contrasts with :class:`ExternalDataTrainingLossMixin`).
    """
    def training_loss(self) -> tf.Tensor:
        """
        The training loss for this model, as used in :class:`gpflow.optimizers.Scipy` and subclasses
        of `tf.optimizers.Optimizer`.
        """
        ...
```
and similar for `ExternalDataTrainingLossMixin`

Reply:
actually that's not quite right, and has just highlighted to me that this PR's doing two things:

  - creating a *uniform API* for training loss across all models (to the extent that's possible). This is `training_loss`
  - adding a *convenience method* that constructs the closure for passing into optimizers. This is `training_loss_closure`

Reply:
```python3
class InternalDataTrainingLossMixin:
    """
    Mixin for models that own their own data. It provides

      - a uniform API for the training loss :meth:`training_loss`
      - a convenience method for constructing the closure expected by various optimizers, namely
        :class:`gpflow.optimizers.Scipy` and subclasses of `tf.optimizers.Optimizer`.

    See :class:`ExternalDataTrainingLossMixin` for an equivalent mixin for models that do **not**
    own their own data.
    """
    def training_loss(self) -> tf.Tensor:
        """ Return the training loss for this model. """
        ...

    def training_loss_closure(self, jit: bool = True) -> Callable[[], tf.Tensor]:
        """
        Convenience method. Returns a closure which can be passed to
        :class:`gpflow.optimizers.Scipy` and subclasses of `tf.optimizers.Optimizer`

        :param jit: If `True`, compile the training loss function in a TensorFlow graph.
        """
```

Reply:
thanks. done

Reply:
and all that documentation made the patch-coverage test pass :sweat_smile: